src/io/read_tree_genesis_hdf5.c

```diff
@@ -0,0 +1,1091 @@
+        
+        //Same with 'Tail' -> 'FirstProgenitor'
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, tail_enum, snap_offset, nhalos_snap, buffer);
+        //the last parameter is '1' for mergertree indices and 0 otherwise
+        ASSIGN_BUFFER_WITH_MERGERTREE_TO_SAGE(nhalos_snap, int64_t, FirstProgenitor, isnap, 1);
+
+        //And same with 'hostHaloID' -> FirstHaloinFOFGroup.
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, hosthaloid_enum, snap_offset, nhalos_snap, buffer);
+        //the last parameter is '1' for mergertree indices and 0 otherwise
+        ASSIGN_BUFFER_WITH_MERGERTREE_TO_SAGE(nhalos_snap, int64_t, FirstHaloInFOFgroup, isnap, 0);
+        
+        /* MS 3rd June, 2019: The LHaloTree convention (which sage uses) is that Mvir contains M200c. While this is DEEPLY confusing,
+           I am using C 'unions' to reduce the confusion slightly. What will happen here is that 'Mass_200crit' will get
+           assigned to the 'M200c' field within the halo struct; but that 'M200c' will also be accessible via 'Mvir'.
+        */
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, m200c_enum, snap_offset, nhalos_snap, buffer);
+        ASSIGN_BUFFER_TO_SAGE(nhalos_snap, double, M200c, float);//M200c is an alias for Mvir
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, vmax_enum, snap_offset, nhalos_snap, buffer);
+        ASSIGN_BUFFER_TO_SAGE(nhalos_snap, double, Vmax, float);
+        
+        /* Read in the positions for the halo centre*/
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, xc_enum, snap_offset, nhalos_snap, buffer);
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, yc_enum, snap_offset, nhalos_snap, buffer + nhalos_snap[0]*sizeof(double));
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, zc_enum, snap_offset, nhalos_snap, buffer + 2*nhalos_snap[0]*sizeof(double));
+        /* Assign to the Pos array within sage*/
+        ASSIGN_BUFFER_TO_NDIM_SAGE(nhalos_snap, NDIM, double, Pos, float);/* Be careful that the last argument is only used
+                                                                             at the assignment stage (i.e., for type conversion)
+                                                                             Otherwise, the memory areas would be accessed incorrectly
+                                                                             and we would get corrupted results -- MS 27/11/2019 */
+        
+        /* Read in the halo velocities */
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, vxc_enum, snap_offset, nhalos_snap, buffer);
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, vyc_enum, snap_offset, nhalos_snap, buffer + nhalos_snap[0]*sizeof(double));
+        READ_PARTIAL_1D_ARRAY(h5_grp, isnap, vzc_enum, snap_offset, nhalos_snap, buffer + 2*nhalos_snap[0]*sizeof(double));
+        /* Assign to the appropriate vel array within sage*/
+        ASSIGN_BUFFER_TO_NDIM_SAGE(nhalos_snap, NDIM, double, Vel, float);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Casting between integer* and double* which have an incompatible binary data representation.](https://app.codacy.com/app/jacobseiler/sage-model/pullRequest?prid=4620074)